File: Data Structures & Algorithms/validate-parentheses/submission-1.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> open;

        for(char c : s){
            if(c == '(' || c == '{' || c == '[') open.push(c);

            else{
                if(open.empty()) return false;

                switch(c){
                    case ')':
                        if(open.top() != '(') return false;
                        else open.pop();
                        break;
                    case '}':
                        if(open.top() != '{') return false;
                        else open.pop();
                        break;
                    case ']':
                        if(open.top() != '[') return false;
                        else open.pop();
                        break;
                }
            }
        }

        return open.empty();
    }
};

File: Data Structures & Algorithms/minimum-stack/submission-1.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minStack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.minStack or self.minStack[-1] >= val:
            self.minStack.append(val)

    def pop(self) -> None:
        if self.minStack and self.stack and self.minStack[-1] == self.stack[-1]:
            self.minStack.pop()
        self.stack.pop()

    def top(self) -> int:
        return None if not self.stack else self.stack[-1]

    def getMin(self) -> int:
        return None if not self.minStack else self.minStack[-1]
